compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -611,14 +611,39 @@ impl<'a, 'tcx> CastCheck<'tcx> {
-                        // When casting a raw pointer to another raw pointer, we cannot convert the cast into
-                        // a coercion because the pointee types might only differ in regions, which HIR typeck
-                        // cannot distinguish. This would cause us to erroneously discard a cast which will
-                        // lead to a borrowck error like #113257.
-                        // We still did a coercion above to unify inference variables for `ptr as _` casts.
-                        // This does cause us to miss some trivial casts in the trival cast lint.
-                        debug!(" -> PointerCast");
+                    if let ty::RawPtr(src_pointee) = self.expr_ty.kind()
+                        && let ty::RawPtr(tgt_pointee) = self.cast_ty.kind()
+                    {
+                        if let Ok(Some(src_kind)) = fcx.pointer_kind(src_pointee.ty, self.expr_span)
+                            && let Ok(Some(tgt_kind)) =
+                                fcx.pointer_kind(tgt_pointee.ty, self.cast_span)
+                        {
+                            match (src_kind, tgt_kind) {
+                                // When casting a raw pointer to another raw pointer, we cannot convert the cast into
+                                // a coercion because the pointee types might only differ in regions, which HIR typeck
+                                // cannot distinguish. This would cause us to erroneously discard a cast which will
+                                // lead to a borrowck error like #113257.
+                                // We still did a coercion above to unify inference variables for `ptr as _` casts.
+                                // This does cause us to miss some trivial casts in the trivial cast lint.
+                                (PointerKind::Thin, PointerKind::Thin)
+                                | (PointerKind::Length, PointerKind::Length) => {
+                                    debug!(" -> PointerCast");
+                                }
+
+                                // If we are not casting pointers to sized types or slice-ish DSTs
+                                // (handled above), we need to make a coercion cast. This prevents
+                                // casts like `*const dyn Trait<'a> -> *const dyn Trait<'b>` which
+                                // are unsound.
+                                //
+                                // See <https://github.com/rust-lang/rust/issues/120217>
+                                (_, _) => {
+                                    debug!(" -> CoercionCast");
+                                    fcx.typeck_results
+                                        .borrow_mut()
+                                        .set_coercion_cast(self.expr.hir_id.local_id);
+                                }
+                            }
+                        }
```

Comment:
This will only work for pointers that can be coerced modulo regions, it will still let something like this pass:

```rust
trait Trait<'a> {}

fn cast<'a, 'b>(x: *mut dyn Trait<'a>) -> *mut (dyn Trait<'b> + Send) {
    x as _
}
```

I think the correct fix would be to actually borrow check the principals for equality around here: https://github.com/rust-lang/rust/blob/d5fd0997291ca0135401a39dff25c8a9c13b8961/compiler/rustc_borrowck/src/type_check/mod.rs#L2294

Also, extending the trait object lifetime itself (e.g. `*mut dyn Trait<'a> + 'a` -> `*mut dyn Trait<'a> + 'static`) is harmless and should be allowed, just like adding auto-traits is.

Reply:
> something like this pass:

It does indeed let it pass, but why? These are [not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6bac84ea7246d552fb022752f049af3b) coercible?..

> I think the correct fix would be to actually borrow check the principals for equality around here: 

I also thought of that, but it seemed somewhat painful to do. And since we handle upcasts with `CastKind::PointerCoercion`, this seemed like a good solution. Borrowck change might be the right solution after all though.

> Edit: borrowck version here: https://github.com/rust-lang/rust/commit/c471b9bfd811c77577e276fb84ddca95c6a018ad

This won't help with `*mut (u8, dyn Trait<'a>) -> *mut (u8, dyn Trait<'b>)`, but that should be an easy fix with `struct_tail`...

> Also, extending the trait object lifetime itself (e.g. *mut dyn Trait<'a> + 'a -> *mut dyn Trait<'a> + 'static) is harmless and should be allowed, [...]

Hm, I can't find an example why it _isn't_ harmless, but I also can't really convince myself that it is. Is this something to do with the fact that you can given `dyn Trait<'a> + 'b` you can use `'a` in `Trait`'s method's signatures, but not `'b`?

> [...], harmless and should be allowed, just like adding auto-traits is.

Adding auto-traits is **not** harmless. Here is an example that segfaults in safe code with arbitraty self receivers and adding send:

<details><summary>example</summary>
<p>

```rust
#![feature(arbitrary_self_types)]

trait Trait {
    fn f(self: *const Self)
    where
        Self: Send;
}

impl Trait for *const () {
    fn f(self: *const Self) {
        unreachable!()
    }
}

fn main() {
    let unsend: *const dyn Trait = &(&() as *const ());
    let send_bad: *const (dyn Trait + Send) = unsend as _;
    send_bad.f();
}
```
```
fish: Job 1, './t' terminated by signal SIGSEGV (Address boundary error)
```
([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3fcadb6854c1d7ce17ade697a3186275))


</p>
</details>

Notably, even though we had to implement `f`, it's not actually in a vtable, since you should not be able to call it (because of unsatisfied `Send` bound):
<details><summary>vtable layout</summary>
<p>

```
error: vtable entries for `<*const () as Trait>`: [
           MetadataDropInPlace,
           MetadataSize,
           MetadataAlign,
           Vacant,
       ]
  --> ./t.rs:24:1
   |
24 | trait Trait {
   | ^^^^^^^^^^^
```
</p>
</details>

Reply:
> Adding auto-traits is **not** harmless. Here is an example that segfaults in safe code with arbitraty self receivers and adding send:

Interesting. I was somehow convinced these functions were still codegened and thought it'd be harmless because you can't do anything bad with with sendable a raw pointer alone.

> Is this something to do with the fact that you can given `dyn Trait<'a> + 'b` you can use `'a` in `Trait`'s method's signatures, but not `'b`?

Yeah that was my reasoning as well. You can add bounds like `Self: 'static`, but since the maybe-not-`'static` thing is behind a raw reference you shouldn't be able to do anything bad with it. But that same reasoning was already wrong for auto traits so I'm not sure anymore.

Also, extending the lifetime of trait objects is often used in unsafe code (like the one in std), so forbidding this will probably cause widespread breakage.

> > something like this pass:
> 
> It does indeed let it pass, but why? These are [not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6bac84ea7246d552fb022752f049af3b) coercible?..

Yeah what I'm trying to say is, that because these are not coercible, the `try_coercion_cast` above will fail, which means that the cast will be a `CastKind::PtrToPtr` and not a coercion. And `CastKind::PtrToPtr` is absolutely not borrow checked so any lifetime can be cast to any other lifetime.

Reply:
> This also means that borrow check alone won't be enough and we need to treat these casts more like coercions.

I _think_ that we can check that in `hir_typeck`, the same way we check
```rust
let x: &dyn Trait + Send = &();
let y: &dyn Trait = x;
```
(I _think_ this is done in `hir_typeck`? will have to find out...)

> What I'm trying to say is that because these are not coercible, the `try_coercion_cast` above will fail, which means that the cast will be a `CastKind::PtrToPtr` and not a coercion. And `CastKind::PtrToPtr` is absolutely not borrow checked so any lifetime can be cast to any other lifetime.

Ah, I see. Right.

Do you think moving the check out of `try_coercion_cast`'s `match` would fix the issue (possibly always reporting an error if `try_coercion_cast` fails for `*mut dyn ...`)? Or should I move to your proposed solution with doing things for `PtrToPtr` in borrowck?

Reply:
> Do you think moving the check out of `try_coercion_cast`'s `match` would fix the issue (possibly always reporting an error if `try_coercion_cast` fails for `*mut dyn ...`)?

Where exactly do you want to move it? If this is checked unconditionally even after `do_check`, then all ptr-ptr casts with vtables that can't be coerced will be just be forbidden, right? Like casting `*mut Wrapper<dyn Trait>` <-> `*mut dyn Trait`.

> I _think_ that we can check that in `hir_typeck`, the same way we check
> 
> ```rust
> let x: &dyn Trait + Send = &();
> let y: &dyn Trait = x;
> ```

I think checking this here in `check_ptr_ptr_cast` where we also compare the principals might be fine. Just check that the source includes all auto traits from the target.

And I'm pretty sure we'd still need borrowck on top of that.

Reply:
> Where exactly do you want to move it? If this is checked unconditionally even after `do_check`, then all ptr-ptr casts with vtables that can't be coerced will be just be forbidden, right? Like casting `*mut Wrapper<dyn Trait>` <-> `*mut dyn Trait`.

Hm, right. With this I'm pretty much convinced that this can't work and we should just handle the lifetimes in borrowck's hadnling of `PtrToPtr`.

> I think checking this here in `check_ptr_ptr_cast` where we also compare the principals might be fine. Just check that the source includes all auto traits from the target.

I think we might have to do something a _little_ more sophisticated to handle `*const dyn X` -> `*const dyn X + Send` where `X` is `trait X: Send {}`, but yeah. Doing this in `check_ptr_ptr_cast` makes sense to me.

> And I'm pretty sure we'd still need borrowck for the principal on top of that.

Yes.

Reply:
I've re-implemented the check, I think it should be correct now. Does it look right to you?

Reply:
`*const dyn Sub` -> `*const Wrapper<dyn Super>` being a "reinterpret cast" is very bad, for the same reasons we have the issues this PR attempts to fix — having an incorrect vtable is not sound with other features. I'll add back the check that the principal traits are the same. While I don't see a safe way to go `*const W<dyn Trait>` -> `*const dyn Trait`, this transformation looks more than fine for me, so I really don't want to depend on it being impossible. (tbh I assumed it was possible, considering you can do `*const dyn Trait` -> `*const W<dyn Trait>`, it's very weird we don't allow the opposite...)

> So I'd argue the current behavior is a bug and it should always be allowed to cast `*const Wrapper<dyn Trait>` to `*const dyn Trait`, which would make your implementation incorrect.

In other words: I agree with this.

Although I'm not sure how you can add this in a non-confusing way.
- `*const ()` -> `*const Send` must be unsizing
- `*const W<dyn Trait>` -> `*const dyn Trait` should be ptr-ptr
  - But what if `W<dyn Trait>: Trait`? Currently it's unsizing, do we keep that?

Reply:
> * But what if `W<dyn Trait>: Trait`? Currently it's unsizing, do we keep that?

`*const W<dyn Trait>` -> `*const dyn Trait` can't be an unsizing cast, because `W<dyn Trait>` is not sized and not a trait object. We can't keep the vtable metadata that is already in the fat pointer `*const W<dyn Trait>` and add the metadata of an `impl Trait for W<dyn Trait>` on top of that.

Casting `*const T` to `*const U` should be an unsizing coercion if and only if `T: Unsize<U>`, but `W<dyn Trait>` can never implement `Unsize<dyn Trait>`.

The problem with the current implementation is that it doesn't fully check `W<dyn Trait>: Unsize<dyn Trait>`. It only sees that `W<dyn Trait>: Unsize<dyn Trait>` holds if the *nested obligations* `W<dyn Trait>: Sized` and `W<dyn Trait>: Trait` hold but then directly assumes the cast must be an unsizing coercion and only checks the nested obligations too late.

Reply:
It is possible to do `*const W<dyn Trait>` -> `*const dyn Trait` right now in safe code. You just need to write a small helper function

```rs
impl<T: ?Sized> Wrapper<T> {
    fn get(this: *const Self) -> *const T {
        this as _
    }
}

fn cast(ptr: *const Wrapper<dyn Super>) -> *const dyn Super {
    Wrapper::get(ptr)
}
```